src/Features/Core/Completion/AbstractCompletionService.cs

```diff
@@ -200,22 +286,52 @@ private class ProviderGroup
             return result;
         }
 
-        private static Task<CompletionItemGroup> GetGroupAsync(
-            ICompletionProvider provider,
+        private static async Task<CompletionList> GetCompletionListAsync(
+            CompletionListProvider provider,
             Document documentOpt,
             SourceText text,
             int position,
             CompletionTriggerInfo triggerInfo,
             CancellationToken cancellationToken)
         {
-            return provider is ITextCompletionProvider
-                ? Task.FromResult(((ITextCompletionProvider)provider).GetGroup(text, position, triggerInfo, cancellationToken))
-                : documentOpt != null
-                    ? provider.GetGroupAsync(documentOpt, position, triggerInfo, cancellationToken)
-                    : SpecializedTasks.Default<CompletionItemGroup>();
+            if (provider is TextCompletionProvider)
+            {
+                return ((TextCompletionProvider)provider).GetCompletionList(text, position, triggerInfo, cancellationToken);
+            }
+
+            if (documentOpt != null)
+            {
+                var itemsBuilder = ImmutableArray.CreateBuilder<CompletionItem>();
```

Comment:
Exact same as in test code.  Do we need to actually have this here.  Or do we own the lifetime of CompletionListContext?  If so, can we just move this into its constructor?